### ISSUE TITLE
Support for Dart

### ISSUE THREAD

**Issue by ghost** [geany/geany#371]
It seems like dart is the only "emerging language" that I don't see syntax highlighting for. it would be nice to have it included, as the Dart Editor and Chrome Dev Editor are memory hogs.

**Comment by insinfo** [comment]
Is dart supported now?

**Comment by elextr** [comment]
Nobody has contributed support.

**Comment by M0JXD** [comment]
I wonder if this might be revisited? Dart is only going to get more popular thanks to the Flutter framework. Ubuntu's installer is now a Flutter application.

There are two pending PRs for Dart support: #379 & #3372

**Comment by elextr** [comment]
#379 is way out of date with conflicts, closed
#3372 also has conflicts that need fixing, if "somebody" addressed it possibly it might happen.

**Comment by M0JXD** [comment]
Ahh yes, the infamous "somebody"... 🤣

My main reason for commenting is that when reading #3938 I saw that Dart was dismissed by yourself as a niche language that there is little point supporting.

While I think that was true at the time this issue was raised in 2014, it is quite the opposite case now. Anyone using an Android phone has almost certainly used a Dart/Flutter application as Google uses the framework, and thanks to Canonical it's now an option for Linux desktop development too. It's become a solid production language with deployed apps being used the world over.

[There are several PRs for Scintilla lexer, which might be the actual resolve?](https://github.com/ScintillaOrg/lexilla/pulls) I imagine there is some extra legwork to integrate it into Geany as per what the mentioned PRs implemented?

I have no qualms about attempting this myself, but I'm not familiar with how Geany works under the hood, nor am I very attuned with Dart yet (I'm learning it now and was hoping to be able to use Geany). I'd be more tempted to have a look once the Dart PRs for Lexilla are merged. Maybe @yangsfang would like to revisit his PR?

**Comment by elextr** [comment]
> My main reason for commenting is that when reading https://github.com/geany/geany/issues/3938 I saw that Dart was dismissed by yourself as a niche language that there is little point supporting.

? where, the only thing in #3938 that is called "niche" is Geany [itself]( https://github.com/geany/geany/issues/3938#issuecomment-2318631896).

**Comment by M0JXD** [comment]
My apologies, my phrasing was poor on that. You said: 

"Only one query for dart since 2014 doesn't seem much support for adding yet another built in language. Just because you added it to Lexilla isn't an argument to build a wasted language to Geany 😜 ."

What I'm trying to say is I don't think it would be a wasted language 🙂

**Comment by elextr** [comment]
Ahh, but the emoji indicates that it is not intended as a serious statement.  

But being serious for a moment, there is no panel or individual who judges if a language is worth adding to Geany, it depends on someone with merge permission either wanting the filetype themselves, or volunteering their own time to merge a PR submitted by someone else, which they do not benefit from.  

It would make it easier to get someone to merge a PR for a new filetype that they don't use if some users of the filetype try the PR and contribute that it is safe and useful (within whatever its limits are).  And the PR is kept up to date if other merges cause clashes.

That reduces or even removes the need for testing by a person who does not use the filetype, has no code in that language to test with, and does not know if it is actually doing the correct thing anyway.  If it is judged that enough testing is happening by people who know the filetype it means all that is needed is a merge and that is more likely to happen as it uses much less of the volunteer's time.

Dart seems to have been unfortunate that even though the PR is a data file only, so no code needs compilation, no Dart users tried it and commented that it was useful (until you now) before other changes caused clashes which were not fixed.

**Comment by yangsfang** [comment]
Hi. I can fix the PR #3372 if you'd like. I wasn't sure why it wasn't merged back in 2023. I vaguely recall it was waiting for approval, but I could be wrong. Anyway, I see another PR already. If you don't get any traction I can revive #3372. 

To be clear, the preliminary support is only for syntax highlighting. It doesn't take into account the parenthesis and other source level information unlike Scintilla. But to get full language support such as meaningful autocomplete you would need to attach a language server.

I don't really use geany as my primary Dart editor. I only use it once in a while to view Dart code where I do not need language support. On my machine geany is patched with my changes, and I never update it. That's why I haven't needed to merge the PR urgently.

It is still useful (in my opinion) to have some kind of Dart support in geany, because geany is built by many Linux distros in the main software repo, as opposed to having to add non-free software repoes for VSCode or another editor.

**Comment by M0JXD** [comment]
I don't know if either of you saw but I fixed the merge conflict with #3968.
Turns out it was a very easy fix, given @yangsfang did all the work with filetype it'd be nice to have your PR merged instead as you deserve the credit. If you have a few minutes it'd be great for you to fix it in your fork and I'll close my PR :)

@elextr I understand what you mean now. I think there is probably a bit of a Chicken and Egg issue where if someone installs Geany and sees there is zero Dart support out of the box (not even syntax highlighting) they will probably go find something else before trying to find PRs that implement the language, even if it is easy to add the filetype. Doubly so given that the Dart/Flutter teams encourage using VSCode, IntelliJ and Android Studio. 

As mentioned I'm learning Dart/Flutter now with intent to use it in a project - while Dart/Flutter with the VSCode plugin is very good and what I will likely use most of the time, I'll be sure to use Geany throughout the process and report back and maybe add improvements, so it can be merged with confidence.

**Comment by elextr** [comment]
@M0JXD @yangsfang 

So now can any Dart users try #3973 which includes the Lexilla highlighter.  Of course it needs Geany to be built with the PR included so its harder to test.

There may be a [Windows installer](https://github.com/geany/geany/actions/runs/11167700273) including the PR, @eht16 I didn't find any instructions on the website, but maybe I skimmed too fast?

**Comment by techee** [comment]
@M0JXD @yangsfang I'm definitely interested in some feedback from Dart users. In particular, I'm not quite sure how the build menu should look like and if what I used from https://github.com/geany/geany/pull/3372 is alright:
```ini
[build-menu]
 # %f will be replaced by the complete filename
 # %e will be replaced by the filename without extension
 FT_00_LB=Dart _Analyze File
 FT_00_CM=dart analyze "%f"
 FT_00_WD=
 EX_00_LB=Flutter Run
 EX_00_CM=flutter run
 EX_00_WD=%p
```

**Comment by techee** [comment]
> There may be a [Windows installer](https://github.com/geany/geany/actions/runs/11167700273) including the PR, @eht16 I didn't find any instructions on the website, but maybe I skimmed too fast?

When you click on Details in "Mingw-w64 Build (Windows)" check, you get to

https://github.com/geany/geany/actions/runs/11177167367/job/31072101463?pr=3973

When you delete the `/job/31072101463?pr=3973` part from the URL (I really haven't found a better way), you get to

https://github.com/geany/geany/actions/runs/11177167367

which at the bottom contains "Artifacts" which is the Windows binary.

**Comment by elextr** [comment]
PR -> Actions -> Select the build action -> Artifacts gives the same URL as you create without the fudging :-)

**Comment by techee** [comment]
Ah, right, I never clicked the "Actions" button before ;-)

**Comment by M0JXD** [comment]
Thank you both :)

I'll build the #3973 version this weekend to try it out, I have some code from going through the Flutter tutorials which I can test with Geany, and I can try some simple CLI stuff using pure dart too. 

I think maybe the analyse commands should be changed to use the "flutter analyze" as opposed to the Dart one: https://docs.flutter.dev/reference/flutter-cli 

I can see it's bit of a nuisance to implement because Flutter has it's own tooling on top of the Dart tooling. Given the main use case for Dart is to build Flutter apps I suppose setting the default build commands to use the Flutter CLI makes the most sense.

(Pure Dart tooling for reference: https://dart.dev/tools/dart-tool)

**Comment by insinfo** [comment]
I use pure dart a lot on a daily basis developing backend applications with dart shelf and frontend with ngdart, much more than with Flutter, so for me it would be more useful for dart tools to be the default.

**Comment by elextr** [comment]
Folks folks, how many commands are we talking about?  Can't we all be friends and have both?

**Comment by M0JXD** [comment]
> Folks folks, how many commands are we talking about? Can't we all be friends and have both? 😁

Of course! It was only a suggestion :)

I've built the #3973 version and it's looking good so far!

![Screenshot from 2024-10-05 20-05-46](https://github.com/user-attachments/assets/e80154d1-cbc6-44e0-9fc3-094aba0bc9db)

This is how I've set the build menu; I've tried to keep in line with Geany's usual way of implementing build commands whilst still offering some Flutter support out of the box:

![Screenshot from 2024-10-05 20-31-28](https://github.com/user-attachments/assets/5ae5056a-ac08-4a33-88b4-e1dd1ecf04f3)

It all seems to work well, the "Pure Dart" commands feel akin to the C workflow in Geany, while   
with Flutter I can hot reload from the command line. I haven't had a look at AngularDart but if @insinfo would like different/additional commands I'm sure they could be worked in. Having a wiki entry covering optimal setups for different Dart use cases (ngdart, Flutter etc.) is something that may be useful. 

Thank you all so much!

**Comment by techee** [comment]
@M0JXD Thanks for your input. For the dart config file though one should only define "Dart commands", not the "Independent commands" which are common for all programming languages.

We are slightly limited in the number of command "slots" Geany offers - i.e. 3 for compile commands and 2 for execute commands. I think that since Dart is the main language, it should be dart commands by default (for execute the second option can be flutter though). How about:
```ini
[build-menu]
# %f will be replaced by the complete filename
# %e will be replaced by the filename without extension
FT_00_LB=_Compile
FT_00_CM=dart compile exe "%f"
FT_00_WD=
FT_01_LB=_Analyze
FT_01_CM=dart analyze "%f"
FT_01_WD=
FT_02_LB=_Test
FT_02_CM=dart test "%f"
FT_02_WD=
EX_00_LB=Dart _Run
EX_00_CM=dart run "%f"
EX_00_WD=%p
EX_01_LB=Flutter R_un
EX_01_CM=flutter run "%f"
EX_01_WD=%p
```

I think the format command you used isn't a good candidate for compile commands as it can modify the current (possibly unsaved) buffer that may need to be reloaded. Dart LSP server is a much better candidate for this.

Speaking of LSP, I've added some default config option for the Dart LSP server to https://github.com/techee/geany-lsp which, if interested, you can try together with https://github.com/geany/geany/pull/3973. As far as I can tell everything works but I'd be interested in real Dart user feedback.

**Comment by techee** [comment]
Fixed by https://github.com/geany/geany/pull/3973.

**Comment by techee** [comment]
> Here is how it looks in VS Code in the context of Dart/Flutter - the Code Actions (under the refactor menu) allow quickly adding code for Flutters "widgets":

I haven't tried the Dart LSP server (I'm on arm64 linux under macOS and haven't found the corresponding dart binaries) but in general this feature should be implemented. When you right-click something in the editor, these should be present under the Commands submenu. There's also a new keybinding you can assign to quickly pop up a menu with all the available commands.

> I haven't had a chance to recompile with the latest support and try your LSP plugin

Would be useful if someone did to see if everything in geany-lsp works for Dart.

**Comment by M0JXD** [comment]
Sorry I missed that, I scanned your readme and because I didn't see "code actions" or similar I presumed it wasn't there.

Anyways I've just tested and it works perfectly! I used the default config with no modification.

![Screenshot from 2024-11-17 21-55-32](https://github.com/user-attachments/assets/d04b4f05-d5dd-4e01-a2e7-2e25231907d5)
![Screenshot from 2024-11-17 21-56-10](https://github.com/user-attachments/assets/58ff86b7-50e9-4687-afd7-d91f91d21e8f)

> I'm on arm64 linux under macOS and haven't found the corresponding dart binaries

This is probably an issue worth raising to the Dart/Flutter teams, ARM is slowly beginning to proliferate the desktop market so they'll have to consider offering these binaries.
